app/utility/base_knowledge_svc.py

```diff
@@ -73,7 +74,18 @@ async def _get_meta_facts(self, meta_fact=None, agent=None, group=None):
     async def _get_fact_origin(self, fact):
         # Retrieve the specific origin of a fact. If it was learned in the current operation, parse through links to
         # identify the host it was discovered on.
-        raise NotImplementedError
+        tempFact = fact # Used so we don't overwrite the supplied fact
+        if getattr(fact, 'links', False) == False:  # Check if the fact we recieved was a fact object or a dict
+            tempFact = await self._get_facts(fact)  # If it was a dictionary, we use the data supplied to search and give us facts that match the criteria
+            if tempFact:    # Make sure we recieved non-empty list
+                tempFact = tempFact[0]  # Select the first fact in the list of facts
+
+        if tempFact.links:      # Check the links of the fact object we have to see if any exist
+            return str(tempFact.links[0].host)  # Return the host that the first link was made on
+        elif tempFact.source:   # If we don't have links, but a source is found for the fact
+            return str(tempFact.source) # Return the source of the fact
+
+        return None # If all the above fails, return none. 
```

Comment:
please take a pass through this code and remove as many comments as possible. where you feel that a comment can't be removed move the code block to a function with a doc string / give the function a descriptive name